crates/bevy_ecs/src/query/iter.rs

```diff
@@ -658,3 +671,66 @@ impl<'w, 's, Q: WorldQuery, F: ReadOnlyWorldQuery> QueryIterationCursor<'w, 's,
         }
     }
 }
+
+// A compile-time checked union of two different types that differs based
+// whether a fetch is dense or not.
+union QuerySwitch<Q, F, A, B> {
+    dense: ManuallyDrop<A>,
+    sparse: ManuallyDrop<B>,
+    marker: PhantomData<(Q, F)>,
+}
+
+impl<Q: WorldQuery, F: WorldQuery, A, B> QuerySwitch<Q, F, A, B> {
+    const IS_DENSE: bool = Q::IS_DENSE && F::IS_DENSE;
```

Comment:
This is a useful rule to explain in the doc strings.

Reply:
This is just mirroring the existing IS_DENSE fields on queries and WorldQueries, should I just point readers there?

Reply:
Yes please